File: src/algorithms/astar.rs
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::cmp::Ordering;

// Struktur untuk merepresentasikan sebuah node dalam graf
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
struct Node {
    id: usize,
    cost: usize,
    heuristic: usize,
}

// Implementasi Ord untuk Node agar dapat digunakan dalam BinaryHeap
impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        // Perbandingan terbalik agar BinaryHeap menjadi min-heap
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

// Fungsi untuk menghitung jarak heuristik (contoh: jarak Manhattan)
fn heuristic(a: (isize, isize), b: (isize, isize)) -> usize {
    ((a.0 - b.0).abs() + (a.1 - b.1).abs()) as usize
}

// Algoritma A*
pub fn astar(
    graph: &HashMap<usize, Vec<(usize, usize)>>,
    start: usize,
    end: usize,
    positions: &HashMap<usize, (isize, isize)>,
) -> Option<(Vec<usize>, usize)> {
    let mut dist: HashMap<usize, usize> = HashMap::new();
    let mut pq = BinaryHeap::new();
    let mut prev: HashMap<usize, usize> = HashMap::new();

    dist.insert(start, 0);
    pq.push(Node {
        id: start,
        cost: 0 + heuristic(positions[&start], positions[&end]),
        heuristic: heuristic(positions[&start], positions[&end]),
    });

    while let Some(Node { id, cost, .. }) = pq.pop() {
        if id == end {
            let mut path = Vec::new();
            let mut curr = end;
            while let Some(&p) = prev.get(&curr) {
                path.push(curr);
                curr = p;
            }
            path.push(start);
            path.reverse();
            return Some((path, dist[&end]));
        }

        if cost > dist.get(&id).cloned().unwrap_or(usize::MAX) {
            continue;
        }

        if let Some(neighbors) = graph.get(&id) {
            for &(neighbor, weight) in neighbors {
                let new_dist = dist[&id] + weight;
                if new_dist < dist.get(&neighbor).cloned().unwrap_or(usize::MAX) {
                    dist.insert(neighbor, new_dist);
                    prev.insert(neighbor, id);
                    pq.push(Node {
                        id: neighbor,
                        cost: new_dist + heuristic(positions[&neighbor], positions[&end]),
                        heuristic: heuristic(positions[&neighbor], positions[&end]),
                    });
                }
            }
        }
    }

    None
}

File: src/algorithms/astar.py
import heapq
from typing import Dict, List, Tuple, Optional
from ..heuristics.euclidean import euclidean_distance
from ..heuristics.manhattan import manhattan_distance


def astar(
    graph: Dict[int, List[Tuple[int, int]]],
    start: int,
    end: int,
    positions: Dict[int, Tuple[float, float]],
    heuristic_type: str = "euclidean"
) -> Optional[Tuple[List[int], int]]:
    # Pilih fungsi heuristik
    if heuristic_type == "manhattan":
        heuristic = manhattan_distance
    else:
        heuristic = euclidean_distance
    
    # Inisialisasi
    dist: Dict[int, int] = {}
    pq = []  # Priority queue: (f_cost, g_cost, node_id)
    prev: Dict[int, int] = {}
    visited = set()
    
    # Inisialisasi node awal
    dist[start] = 0
    h_start = heuristic(positions[start], positions[end])
    heapq.heappush(pq, (h_start, 0, start))
    
    while pq:
        f_cost, g_cost, node_id = heapq.heappop(pq)
        
        # Skip jika sudah dikunjungi dengan cost lebih baik
        if node_id in visited:
            continue
        
        visited.add(node_id)
        
        # Jika mencapai tujuan
        if node_id == end:
            # Rekonstruksi jalur
            path = []
            curr = end
            while curr is not None:
                path.append(curr)
                curr = prev.get(curr)
            path.reverse()
            return (path, dist[end])
        
        # Periksa semua tetangga
        if node_id in graph:
            for neighbor, weight in graph[node_id]:
                if neighbor in visited:
                    continue
                
                new_g_cost = dist[node_id] + weight
                
                # Jika menemukan jalur yang lebih baik
                if neighbor not in dist or new_g_cost < dist[neighbor]:
                    dist[neighbor] = new_g_cost
                    prev[neighbor] = node_id
                    h_cost = heuristic(positions[neighbor], positions[end])
                    f_cost = new_g_cost + h_cost
                    heapq.heappush(pq, (f_cost, new_g_cost, neighbor))
    
    return None

